TOF symmetries rework

robbietuk opened the pull request:
This PR reworks a lot of #9 by building upon @NikEfth's initial cache work.

I went with the approach that the LORs (a.k.a probability elements) on the cartesian grid should be correct on application of the current symmetries. Then adjust the `basic_bin` -> `bin` convertions. This essentially entails swapping `tof_pos_num` with the enforcement of `do_symmetry_swap_s && tangential_pos_num < 0` in `find_basic_bin`.

 - SymmetryOperation_PET_CartesianGrid classes use now use `axial_pos_shift = 0` if `do_symmetry_shift_z`. This was needed becasue the axial position was is not set to 0 in the `basic_bin` but SymmetryOperations were adding the shift anyway. I believe this was discovered due to extended tests. 
 - Reworked a lot of the `apply_tof_kernel` to simplify the methodology and remove symmetry operations from the method.
 - Rework a lot of the `cache_stores_only_basic_bins` and `!cache_stores_only_basic_bins` cases to deal with caching in a simpler manner.
 - Fix `SymmetryOperation_PET_CartesianGrid_swap_zq:: transform_proj_matrix_elems_for_one_bin` (see https://github.com/UCL/STIR/issues/995) because symmetries use the method. This will probably introduce a merge issue with UCL/STIR but lets fix later.
 - Improve upon the current symmetry tests by adding other symmetry configurations and TOF scanner.

NikEfth committed:
Use Bin TOF position in cache

In addition remove elements from LORs instead settings them to zero.

robbietuk committed:
Rework get_proj_matrix_elems_for_one_bin and simply apply_tof_kernel

robbietuk committed:
find_basic_bin `-1 * timing_pos_num` if  do_symmetry_swap_s && tangential_pos_num < 0

robbietuk committed:
Add SymmetryOperation_PET_CartesianGrid_swap_xmx_ymy_tmt

robbietuk committed:
Improve test by inspecting elems_with_sym.bin is same as bin

robbietuk committed:
Changes to DataSymmetries, if shift_z, set axial_pos_shift to axial pos, otherwise 0

Also delete previous tof_pos_num parsing.

robbietuk committed:
Fix timing_pos_num changes for SymOp swap_xmx_ymy_zq

robbietuk committed:
Add more tests and fix TOF scanner config

robbietuk commented on src/recon_test/test_DataSymmetriesForBins_PET_CartesianGrid.cxx:
```diff
@@ -124,14 +124,22 @@ run_tests_2_proj_matrices_1_bin(const ProjMatrixByBin& proj_matrix_no_symm,
 #endif
       elems_no_sym.sort();
       elems_with_sym.sort();
-      if (!check(elems_no_sym == elems_with_sym, "comparing lors"))
+
```
Getting an error that these sorts should be preventing... Note the coordinates `{2, 0, 0}` and `{2, 1, 0}` appear twice, once for each `elems`. Note, these results are with a scanner with 0 phi offset.
```
Error. Comparing symmetry LORs
	bin: 	segment = -3, 	axial pos 0, 	view = 0, 	tangential_pos_num = 0, 	timing position index = -1
SymBin :	segment = -3, 	axial pos 0, 	view = 0, 	tangential_pos_num = 0, 	timing position index = -1
{2, 0, 0}:0.000120942    ||   
{2, 1, 0}:0.000322843    ||   
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
CUT TO SIMPLIFY
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
{4, -28, 1}:5.42422e-09    ||   {4, -28, 1}:5.70331e-09
{4, -27, 0}:2.05508e-09    ||   {4, -27, 0}:2.159e-09
{4, -27, 1}:6.63398e-09    ||   {4, -27, 1}:6.97077e-09
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
CUT TO SIMPLIFY
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
                       ||   {3, 32, -1}:0.125781
                       ||   {2, 0, 0}:0.000124743
                       ||   {2, 1, 0}:0.000332781
                       ||   {2, 2, 0}:0.000441099
                       ||   {2, 3, 0}:0.000581024
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
CUT TO SIMPLIFY
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
```

robbietuk reviewed (commented):
Locally, the `test_DataSymmetriesForBins_PET_CartesianGrid` is passing. However, I just realised `WARNING: Disabling symmetries as image is rotated due to phi offset of the scanner.` is occuring. I am adding a test of an additional scanner with no phi offset to reenable 180 and 90 degree symmetries and test those. Likely this is not done.

robbietuk committed:
Turn off data symmetries if tof enabled

also check if do_symmetry_90degrees_min_phi and do_symmetry_180degrees_min_phi already off before turning them off repeatedly. cleaner errors

robbietuk commented:
Rebased onto `tof_sino_UCL`

Observed issue with larger scanners (not committed), possibly related to https://github.com/UCL/STIR/issues/1002, where symmetry values are slightly larger than non-symmetry when `tangential_pos_num >= 0`, likely poorly selected symmetries?

Current GHA error is in `test_time_of_flight` (and #26 which I am ignoring for now)
```
Error : unequal values are 0.50209 and 0.291119. Sum over nonTOF LOR does not match sum over TOF LOR.
```

Does not seem related to the above error as it exists when the `test_time_of_flight` scanner symmetries are disabled.

robbietuk commented:
> Observed issue with larger scanners (not committed), possibly related to https://github.com/UCL/STIR/issues/1002, where symmetry values are slightly larger than non-symmetry when tangential_pos_num >= 0, likely poorly selected symmetries?

Investigating this further, I find the errors are only occur in the combination of `segment != 0`, `timing position index != 0`, and `tangential_pos_num >= 0`. The error does occur at the 0th view and all view symmetries are false. The image is `(19, 19, 6)` (x,y,z). This error occurs with and without the `erf` changes. 

Maybe this is being caused by the application of the TOF kernal onto the `basic_bin` and then a symmetry applied. This likely introduces an inconsistency when finding centers of voxels etc. However, that would not explain the `tangential_pos_num >= 0` cases.

Example error is typical and there is less %error at lower segment magnetude.
```
Current bin: 	segment = -3, 	axial pos 0, 	view = 0, 	tangential_pos_num = 0, timing position index = -1
Symm bin: 	segment = -3, 	axial pos 0, 	view = 0, 	tangential_pos_num = 0, timing position index = -1
No Symmetries Iterator || Symmetries Iterator 
      no_sym_iter         ||        sym_iter           ||     Error       
{2, 0, 0}:0.000120939    ||   {2, 0, 0}:0.000124746     ||    abs(-3.05202%) > 1%
{2, 1, 0}:0.000322844    ||   {2, 1, 0}:0.000332781     ||    abs(-2.98631%) > 1%
{2, 2, 0}:0.000428212    ||   {2, 2, 0}:0.000441099     ||    abs(-2.9215%) > 1%
{2, 3, 0}:0.000564424    ||   {2, 3, 0}:0.000581025     ||    abs(-2.85716%) > 1%
{2, 4, 0}:0.000739332    ||   {2, 4, 0}:0.000760571     ||    abs(-2.79251%) > 1%
{2, 5, 0}:0.000962423    ||   {2, 5, 0}:0.000989418     ||    abs(-2.72834%) > 1%
{2, 6, 0}:0.00124507    ||   {2, 6, 0}:0.00127915     ||    abs(-2.66414%) > 1%
{2, 7, 0}:0.00160079    ||   {2, 7, 0}:0.00164353     ||    abs(-2.60016%) > 1%
{2, 8, 0}:0.00204547    ||   {2, 8, 0}:0.0020987     ||    abs(-2.53626%) > 1%
{2, 9, 0}:0.00259764    ||   {2, 9, 0}:0.0026635     ||    abs(-2.47268%) > 1%
{3, -9, 0}:1.34483e-05    ||   {3, -9, 0}:1.39562e-05     ||    abs(-3.63954%) > 1%
{3, -8, 0}:1.9007e-05    ||   {3, -8, 0}:1.97114e-05     ||    abs(-3.574%) > 1%
{3, -7, 0}:2.66921e-05    ||   {3, -7, 0}:2.76627e-05     ||    abs(-3.50844%) > 1%
{3, -6, 0}:3.72462e-05    ||   {3, -6, 0}:3.85743e-05     ||    abs(-3.44303%) > 1%
{3, -5, 0}:5.16433e-05    ||   {3, -5, 0}:5.34486e-05     ||    abs(-3.37757%) > 1%
{3, -4, 0}:7.11515e-05    ||   {3, -4, 0}:7.3589e-05     ||    abs(-3.31231%) > 1%
{3, -3, 0}:9.74087e-05    ||   {3, -3, 0}:0.000100678     ||    abs(-3.24713%) > 1%
{3, -2, 0}:0.000132513    ||   {3, -2, 0}:0.000136869     ||    abs(-3.18215%) > 1%
{3, -1, 0}:0.000179133    ||   {3, -1, 0}:0.000184896     ||    abs(-3.11703%) > 1%
{3, 0, 0}:0.000240631    ||   {3, 0, 0}:0.000248207     ||    abs(-3.05237%) > 1%
{3, 1, 0}:0.000321214    ||   {3, 1, 0}:0.000331106     ||    abs(-2.98733%) > 1%
{3, 2, 0}:0.000426098    ||   {3, 2, 0}:0.000438927     ||    abs(-2.92276%) > 1%
{3, 3, 0}:0.0005617    ||   {3, 3, 0}:0.000578226     ||    abs(-2.8581%) > 1%
{3, 4, 0}:0.000735844    ||   {3, 4, 0}:0.000756992     ||    abs(-2.79365%) > 1%
{3, 5, 0}:0.00095799    ||   {3, 5, 0}:0.000984871     ||    abs(-2.72936%) > 1%
{3, 6, 0}:0.00123948    ||   {3, 6, 0}:0.00127342     ||    abs(-2.66528%) > 1%
{3, 7, 0}:0.00159377    ||   {3, 7, 0}:0.00163633     ||    abs(-2.6013%) > 1%
{3, 8, 0}:0.00203672    ||   {3, 8, 0}:0.00208974     ||    abs(-2.53748%) > 1%
{3, 9, 0}:0.00258681    ||   {3, 9, 0}:0.00265242     ||    abs(-2.47375%) > 1%
{4, -9, 0}:1.33654e-05    ||   {4, -9, 0}:1.38703e-05     ||    abs(-3.64069%) > 1%
{4, -8, 0}:1.88919e-05    ||   {4, -8, 0}:1.95923e-05     ||    abs(-3.57502%) > 1%
{4, -7, 0}:2.65335e-05    ||   {4, -7, 0}:2.74986e-05     ||    abs(-3.50959%) > 1%
{4, -6, 0}:3.70291e-05    ||   {4, -6, 0}:3.83499e-05     ||    abs(-3.44407%) > 1%
{4, -5, 0}:5.1348e-05    ||   {4, -5, 0}:5.31436e-05     ||    abs(-3.37886%) > 1%
{4, -4, 0}:7.07526e-05    ||   {4, -4, 0}:7.31773e-05     ||    abs(-3.31345%) > 1%
{4, -3, 0}:9.68734e-05    ||   {4, -3, 0}:0.000100126     ||    abs(-3.24854%) > 1%
{4, -2, 0}:0.0001318    ||   {4, -2, 0}:0.000136134     ||    abs(-3.18318%) > 1%
{4, -1, 0}:0.000178189    ||   {4, -1, 0}:0.000183924     ||    abs(-3.1183%) > 1%
{4, 0, 0}:0.000119695    ||   {4, 0, 0}:0.000123464     ||    abs(-3.05247%) > 1%
```

robbietuk committed:
[ci skip] Disabled all but shift_z, improve test logic, add new scanner

robbietuk commented:
Enabling `do_symmetry_shift_z` as the only symmetry, this error is removed. Althought `test_time_of_flight` still fails. `swap_segments` and `shift_s` may be too hard for TOF right now.

robbietuk committed:
Remove error on non-cached forward projector with TOF

robbietuk committed:
[ci skip] Cleanup and remove tmt symmetry class

robbietuk committed:
Simplify apply_tof_kernal loop

robbietuk commented:
All tests should now pass but `test_time_of_flight` because of #26 
```
Error. get_det_pos_for_bin with bins of opposite timing_pos
```

Check logs and likely ready to merge?

robbietuk commented on src/include/stir/recon_buildblock/DataSymmetriesForBins_PET_CartesianGrid.inl:
```diff
@@ -263,7 +263,7 @@ find_sym_op_general_bin(
                         int view_num, 
                         int axial_pos_num) const
-{
+{ 
```
```suggestion
{
```

robbietuk commented on src/include/stir/recon_buildblock/DataSymmetriesForBins_PET_CartesianGrid.inl:
```diff
@@ -387,7 +387,8 @@ find_sym_op_general_bin(
       else // segment_num = 0 
       {
-        if ( do_symmetry_swap_s && s < 0) return new SymmetryOperation_PET_CartesianGrid_swap_xmx_ymy(view180, axial_pos_shift, z_shift);
+        if ( do_symmetry_swap_s && s < 0)
+          return new SymmetryOperation_PET_CartesianGrid_swap_xmx_ymy(view180, axial_pos_shift, z_shift);
```
```suggestion
        if ( do_symmetry_swap_s && s < 0) return new SymmetryOperation_PET_CartesianGrid_swap_xmx_ymy(view180, axial_pos_shift, z_shift);
```

robbietuk commented on src/include/stir/recon_buildblock/DataSymmetriesForBins_PET_CartesianGrid.inl:
```diff
@@ -477,9 +478,9 @@ find_basic_view_segment_numbers(ViewSegmentNumbers& v_s) const
 }
 
-bool  
+bool
```
```suggestion
bool  
```

robbietuk commented on src/recon_buildblock/SymmetryOperations_PET_CartesianGrid.cxx:
```diff
@@ -635,6 +635,7 @@ transform_proj_matrix_elems_for_one_densel(
 } 
 
 
+
```
```suggestion
```

robbietuk reviewed (commented):
White space change suggestions

robbietuk committed:
[ci skip] white space fixes

robbietuk commented:
Looks good. GHA only error with 
>Error. get_det_pos_for_bin with bins of opposite timing_pos

https://github.com/NikEfth/STIR/runs/5401057535?check_suite_focus=true#step:8:118, the same as `tof_sino_ucl`.

KrisThielemans commented on src/recon_buildblock/ProjMatrixByBin.cxx:
```diff
@@ -171,19 +171,23 @@ set_up(
-    <li>abs(tangential_pos_num) fits in 11 bits</li>   
+    <li>abs(axial_pos_num) fits in 13 (4095) bits
+    <li>abs(tangential_pos_num) fits in 10 (1024) bits
+    <li>abs(tof_pos_num) fits in 7 bits (127)
```
not enough...

robbietuk replied:
Needs more, but how many more bits? Should we heuristicly set this by setting the value greater than the `timing_pos_numbers` of any scanner.

KrisThielemans replied:
heuristic (who knows how many TOF bins a future scanner will have?). So I've given it plenty now by going to 64 bit

KrisThielemans commented on src/include/stir/recon_buildblock/ProjMatrixByBin.inl:
```diff
@@ -55,124 +54,95 @@ get_proj_matrix_elems_for_one_bin(
-        new_value = get_tof_value(low_dist, high_dist);
-        new_value *= element_ptr->get_value();
-        *element_ptr = ProjMatrixElemsForOneBin::value_type(c, new_value);
+        if ((low_dist >= 4.f && high_dist >= 4.f) ||  (low_dist <= -4.f && high_dist <= -4.f))
```
I feel we should move this `if` inside `get_tof_value`

KrisThielemans replied:
done

KrisThielemans reviewed (changes requested):
a difficult one for the caching...

robbietuk commented on src/recon_buildblock/ProjMatrixByBin.cxx:
```diff
@@ -171,19 +171,23 @@ set_up(
-                    | (static_cast<boost::uint32_t>(abs(bin.axial_pos_num()))<<13) 
-                    | (static_cast<boost::uint32_t>(bin.tangential_pos_num()>=0?0:1) << 12)
-                    |  static_cast<boost::uint32_t>(abs(bin.tangential_pos_num())) );    	
+  assert(static_cast<boost::uint32_t>(abs(bin.axial_pos_num())) < (static_cast<boost::uint32_t>(1)<<13));
+  assert(abs(bin.tangential_pos_num()) < (1<<10));
```
Need a TOF assert? e.g. `assert(abs(bin.timing_pos_num()) < (1<<7));`

The value 7 to change.

robbietuk replied:
Also, both of these should the same type. Keep the `static_cast` or remove it?

KrisThielemans committed:
more bits for ProjMatrixByBin cache_key

KrisThielemans committed:
minor simplification in TOF kernel application

KrisThielemans committed:
comment-out failing TOF det_pos test for now

KrisThielemans committed:
corrected assert [skip appveyor]

KrisThielemans commented:
@robbietuk could you have a quick look at my changes and let me know? Otherwise good to go I believe

robbietuk commented:
Looks good to me.

robbietuk commented on src/recon_buildblock/ProjMatrixByBin.cxx:
```diff
@@ -171,19 +187,24 @@ set_up(
+  assert(static_cast<CacheKey>(abs(bin.timing_pos_num())) < (static_cast<CacheKey>(1) << timing_pos_bits));
+
+  return static_cast<CacheKey>(
+                               (static_cast<CacheKey>(bin.axial_pos_num()>=0?0:1) << (timing_pos_bits + tang_pos_bits + axial_pos_bits + 3))
+                               | (static_cast<CacheKey>(abs(bin.axial_pos_num())) << (timing_pos_bits + tang_pos_bits + 2))
```
Why does this increase by `axial_pos_bits + 1` from the previous?

KrisThielemans replied:
which line? first one? I think you're right that it should be `timing_pos_bits + tang_pos_bits + axial_pos_bits +2`. If that's what you thought as well, please correct it.

I always confuse myself about these things (sadly cannot use a bit field as it doesn't have equality operator). I guess it be better to have a test that they don't overlap (important), or go 1 too far, which this seems to be doing. I couldn't face writing one though...

robbietuk replied:
Fixed. I agree. I think this is now correct. I had a similar WIP as well using variables because it is safer.

robbietuk committed:
Correct bit shift

KrisThielemans commented:
a pity of the `ci skip`. did you run it locally?

robbietuk commented:
Yes, passed tests. Also, minor changes since last tested with GHA.

robbietuk commented:
Merge?

robbietuk commented:
@KrisThielemans Sorry, what is the status on this? I think that it works with the current z-shift symmertries but we haven't enough memory for a full scanner? Running GHA.

robbietuk commented on src/include/stir/recon_buildblock/DataSymmetriesForBins_PET_CartesianGrid.inl:
```diff
@@ -479,7 +479,7 @@ find_basic_view_segment_numbers(ViewSegmentNumbers& v_s) const
 bool  
 DataSymmetriesForBins_PET_CartesianGrid::
-find_basic_bin(int &segment_num, int &view_num, int &axial_pos_num, int &tangential_pos_num) const 
+find_basic_bin(int &segment_num, int &view_num, int &axial_pos_num, int &tangential_pos_num, int &timing_pos_num) const
```
Pass `bin`?

robbietuk replied:
A change to this functionality is out of scope of this PR. `inline bool find_basic_bin(Bin& b) const;` does exist and is a private member but its functionality is fairly specific.

KrisThielemans replied:
sure

robbietuk commented:
Merge `tof_sino_UCL` and run tests. 

Do this PR before #17

robbietuk commented:
I rebased onto `tof_sino_UCL` and all tests pass locally.

KrisThielemans committed:
re-enable test_tof_proj_data_info_det_pos as now fixed

KrisThielemans commented:
This PR seems fine, but when testing it via `recon_test_pack` with TOF (not yet merged), it shows there's a problem in the TOF code for `PoissonLogLikelihooodWithLinearModelFormMeanAndProjData` when using the non-TOF projector for the sensitivity, as symmetries are not the same between TOF and non-TOF. (This wasn't a problem before this PR, as then the symmetries didn't take TOF into account). I'll push relevant test later.

KrisThielemans committed:
Merge remote-tracking branch 'nikos/tof_sino_UCL' into TOF/Symmetries/Main

KrisThielemans commented:
In addition to the above problem with the symmetries, another problem is that `setup_distributable_computation` needs to be called with the correct projector (although this actually only matters when using MPI).

KrisThielemans committed:
sensitivity fixes for the TOF case

Symmetries between the TOF and non-TOF case can now be different, so we need
to take this into account.

In addition, make sure that we call setup_distributable_computation with
the appropriate backprojector (depending if we do the gradient or
sensitivity calculation)
